Added Determinant function in SqaureMat class

ayush9398 opened the pull request:
Calculates determinant for a square matrix.

ayush9398 commented:
@8hantanu if I understand it correctly, the `self.value` contains the matrix in the form of a multi-dimensional array, `mat` is the matrix after processing, taken from the `self.value` array. It worked when I tried it on the python console by importing the structure.py file and set `self.value` as a python array, followed by calling the dimensions function of `SquareMat` object and then the determinant function.
Do I need to make any changes further?

8hantanu commented:
For example if the input give is `[1, 2; 3, 4]` (a 2x2 square matrix), the [tokens list will be generated](https://github.com/aerospaceresearch/visma/blob/master/visma/io/tokenize.py#L1557) for each element and will be updated in the `self.value[i][j]`. 
In this case `self.value[i][j]` will be a list of size 1 with element of class type **_visma.functions.constant.Constant_**.

So your operation `mat[0][0] * mat[1][1]`
must look something like `mat.value[0][0][0].value * mat.value[1][1][0].value`

**Note**: The above can only be the used when all elements in the matrix are of **_Constant_** type.

Also you can test your function on a matrix generated from `getTokens("[1, 2; 3, 4]")` in **tests/tester.py** and if possible you can add a simple testcase in **tests/test_matrix.py** verifying the implemented function. Make sure to use `isSquare()` before calling `determinant()` as the class type will not change to **_SquareMat_** unless `isSquare()` is called.

ayush9398 commented:
Okay, I try to resolve this. Thank you @8hantanu

ayush9398 commented:
@8hantanu I have made some changes. Now the determinant function will give the determinant in the form of an array of tokens since the input matrix can have expressions. Please review it, thank you.

PulkitMishra commented:
Hey @ayush9398 run the command pylama on your system before pushing changes. You can easily check for code style with that.

ayush9398 commented:
@8hantanu I have added a `test_determinant('<your matrix here>')` function in the tests/test_matrix.py file. I have used numpy in order to use it's array concatenate operation as it is better than using the `for` loops, which can be time-consuming for calculating the determinant for large matrices.

mayankDhiman commented:
Hi, @ayush9398! Are you working on this?
I was thinking to implement determinant, inverse and adjoint. Determinant is a pre-requisite for that. Should I hit this one up? :)

ayush9398 commented:
@mayankDhiman yes, I am still working on this, just waiting for @8hantanu to review the changes. Please work on any other issue.

8hantanu commented on tests/test_matrix.py:
```diff
@@ -278,3 +278,11 @@ def test_multiplyMatrix():
+def test_determinant(mat='[1,2;3,4]'):
+    mat = getTokens(mat)
+    if mat.isSquare():
+        return mat.determinant()
```
Use `assert` to verify your function's output.
```suggestion
        assert mat.determinant() == <determinant-value>
```
Also add test cases for matrix of size 1 and 3.
Lines 287-288 are not required.

8hantanu commented on tests/test_matrix.py:
```diff
@@ -278,3 +278,11 @@ def test_multiplyMatrix():
+
+
+def test_determinant(mat='[1,2;3,4]'):
+    mat = getTokens(mat)
```
```suggestion
    mat = getTokens('[1,2;3,4]')
```

ayush9398 replied:
@8hantanu I just checked that getTokens only returns the Binary object and not any specific operator object, while I used the specific operator objects like for Multiply, Minus, etc. So, should I change this in my implementation?

8hantanu replied:
What input are you giving to `getTokens()` ?

ayush9398 replied:
input was `1*4-2*3`

8hantanu replied:
Though the type is _Binary_, the `token.value` will have the correct operator symbol.

8hantanu commented on tests/test_matrix.py:
```diff
@@ -278,3 +278,11 @@ def test_multiplyMatrix():
     pass
+
+
+def test_determinant(mat='[1,2;3,4]'):
```
```suggestion
def test_determinant():
```

ayush9398 commented:
Hey @8hantanu, I have added the changes.

ayush9398 commented:
I tried importing the simplify function in the visma/matrix/structure.py file but it is throwing an error:
>>> from tests.test_matrix import test_determinant
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/naruto/Desktop/visma/tests/test_matrix.py", line 2, in <module>
    from visma.matrix.operations import simplifyMatrix, addMatrix, scalarAdd, scalarSub, scalarMult, scalarDiv
  File "/home/naruto/Desktop/visma/visma/matrix/operations.py", line 4, in <module>
    from visma.simplify.simplify import simplify
  File "/home/naruto/Desktop/visma/visma/simplify/simplify.py", line 17, in <module>
    from visma.io.parser import tokensToString
  File "/home/naruto/Desktop/visma/visma/io/parser.py", line 7, in <module>
    from visma.matrix.structure import Matrix
  File "/home/naruto/Desktop/visma/visma/matrix/structure.py", line 8, in <module>
    from visma.simplify.simplify import simplify
ImportError: cannot import name 'simplify' from 'visma.simplify.simplify' (/home/naruto/Desktop/visma/visma/simplify/simplify.py)

8hantanu commented:
Could be due to a circular import. Try putting the `from visma.simplify.simplify import simplify` statement just before you call the `simplify()` function.

AbishekSeshan commented:
Hi @ayush9398 :)  Is visma constricted to 3x3 matrices only? I took a look at your code and I think it only works for 3x3 matrices. If it needs to work for all matrices, I think determinant() needs to be a recursive function don't you think? Because for a 4x4 matrix, we need to find determinants of 3x3 matrices 4 times.

ayush9398 commented:
Hey @AbishekSeshan the determinant function is a recursive function and does support square matrices even greater than 3x3 dimensions, please have a look at line 158 of the visma/matrix/structure.py file.

AbishekSeshan commented:
> Hey @AbishekSeshan the determinant function is a recursive function and does support square matrices even greater than 3x3 dimensions, please have a look at line 158 of the visma/matrix/structure.py file.

Oh shoot I'm really sorry @ayush9398 !!! Don't know how I missed that!

ayush9398 commented:
no problem :)

ayush9398 commented:
@8hantanu I am trying to implement simplify function at line 168 of the structure.py file:
`ans, _, _, _, _ = simplify(mat[0][0] + [a] + mat[1][1] + [b] + mat[0][1] + [a] + mat[1][0])`
I tried to get determinant for matrix `[1,2;3,4]`, although the list of tokens is correct, the solution comes out to be `12.0` instead of `-2.0`. 
Also, I found that the simplify function solves `1*4-2*3` string correctly through the `getTokens` function.
On looking at the simplify functions' animation list, I found that the error is, `4-2*3` tokens list is parsed as `2*6` which is `(4-2)*(2*3)`.
Could you please suggest to me where I should look for this error?

ayush9398 commented:
Hey @8hantanu I have added the changes in the determinant function and also, updated the test function. I am sorry, it took me more time than expected since I was writing GSOC proposal.
